File: impl/ts/src/sandbox/jail.ts
// Filesystem jail: every relative path from an API client must resolve to a
// location strictly inside the sandbox workdir. Mirrors the Python resolve_in_jail:
// reject absolute paths, drive/UNC roots, and symlink/reparse escapes.

import fs from "node:fs";
import path from "node:path";
import { PathEscape } from "../types.js";

export function resolveInJail(workdir: string, relPath: string | null | undefined): string {
  if (relPath === null || relPath === undefined) {
    throw new PathEscape("Path is required.");
  }
  const raw = String(relPath);

  // Reject absolute paths (POSIX or Windows) and UNC.
  if (path.isAbsolute(raw)) {
    throw new PathEscape(`Absolute paths are not allowed: ${raw}`);
  }
  // Normalize separators and strip any leading slashes.
  const candidate = raw.replace(/\\/g, "/").replace(/^\/+/, "");
  // A drive letter inside the first segment (e.g. "C:") is an absolute escape.
  const firstSeg = candidate.split("/")[0] ?? "";
  if (firstSeg.includes(":")) {
    throw new PathEscape(`Drive-absolute paths are not allowed: ${raw}`);
  }

  const root = path.resolve(workdir);
  const target = path.resolve(root, candidate);

  // Containment check on the lexical path.
  const rel = path.relative(root, target);
  if (rel === ".." || rel.startsWith(".." + path.sep) || path.isAbsolute(rel)) {
    throw new PathEscape(`Path escapes the sandbox: ${raw}`);
  }

  // Symlink/reparse check: the real path must also stay inside the root.
  let real = target;
  try {
    real = fs.realpathSync(target);
  } catch {
    // Target may not exist yet (writes); walk up to the nearest existing parent.
    let probe = path.dirname(target);
    while (probe && probe !== path.dirname(probe)) {
      if (fs.existsSync(probe)) {
        try {
          real = path.join(fs.realpathSync(probe), path.relative(probe, target));
        } catch {
          real = target;
        }
        break;
      }
      probe = path.dirname(probe);
    }
  }
  const realRoot = fs.existsSync(root) ? fs.realpathSync(root) : root;
  const realRel = path.relative(realRoot, real);
  if (realRel === ".." || realRel.startsWith(".." + path.sep) || path.isAbsolute(realRel)) {
    throw new PathEscape(`Path escapes the sandbox via a link: ${raw}`);
  }
  return target;
}

File: impl/go/internal/sandbox/jail.go
// Filesystem jail: resolve a client-supplied relative path strictly inside the
// sandbox workdir. Mirrors the Python resolve_in_jail.
package sandbox

import (
	"path/filepath"
	"strings"

	"optimaize-videomaker-go/internal/types"
)

func resolveInJail(workdir, relPath string) (string, error) {
	if relPath == "" {
		return "", types.PathEscape("Path is required.")
	}
	// Reject absolute paths and drive/UNC roots.
	if filepath.IsAbs(relPath) {
		return "", types.PathEscape("Absolute paths are not allowed: %s", relPath)
	}
	candidate := strings.ReplaceAll(relPath, "\\", "/")
	candidate = strings.TrimLeft(candidate, "/")
	first := candidate
	if idx := strings.IndexByte(candidate, '/'); idx >= 0 {
		first = candidate[:idx]
	}
	if strings.Contains(first, ":") {
		return "", types.PathEscape("Drive-absolute paths are not allowed: %s", relPath)
	}

	root, err := filepath.Abs(workdir)
	if err != nil {
		return "", types.OperationError("bad workdir: %v", err)
	}
	target := filepath.Join(root, filepath.FromSlash(candidate))
	rel, err := filepath.Rel(root, target)
	if err != nil || rel == ".." || strings.HasPrefix(rel, ".."+string(filepath.Separator)) || filepath.IsAbs(rel) {
		return "", types.PathEscape("Path escapes the sandbox: %s", relPath)
	}

	// Symlink/reparse check via EvalSymlinks where the path exists.
	realRoot, err := filepath.EvalSymlinks(root)
	if err != nil {
		realRoot = root
	}
	if real, err := filepath.EvalSymlinks(target); err == nil {
		rr, err2 := filepath.Rel(realRoot, real)
		if err2 != nil || rr == ".." || strings.HasPrefix(rr, ".."+string(filepath.Separator)) || filepath.IsAbs(rr) {
			return "", types.PathEscape("Path escapes the sandbox via a link: %s", relPath)
		}
	}
	return target, nil
}
